platform/lib/templates/SupportedFormatsExtension.js

```diff
@@ -0,0 +1,58 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+
+/**
+ * Helps determining the default format for a certain template
+ * to render it in the best fitting format if none is explicitly set
+ */
+class SupportedFormatsExtension {
+  constructor() {
+    this.tags = ['supportedFormats'];
+  }
+
+  parse(parser, nodes) {
+    const token = parser.nextToken();
+
+    // Parse arguments from tag
+    const args = parser.parseSignature(null, true);
+    parser.advanceAfterBlockEnd(token.value);
+
+    return new nodes.CallExtension(this, 'run', args);
+  }
+
+  /**
+   * Tries to figure out if there is a fitting default
+   * format and overwrites the one set by the global context
+   */
+  run(context, args) {
+    const formats = args.formats || [];
```

Comment:
I don't quite understand why we need an extension for this?

Reply:
We don't need one necessarily. An equivalent solution would be to put the logic into the template:

```
{% if doc.formats %}
[% if not requestedFormat %]
[% set supportedFormats = {{ doc.formats|tojson }} %]
[% set format = supportedFormats[0] || 'websites' %]
[% endif %]
{% endif %}
```